### PR TITLE
Axolotl supports falcon + qlora

### PR DESCRIPTION
**Pull request by utensil**
This PR:

- [x] confirms that axolotl supports falcon + qlora 
- [x] adds `example/falcon/config-7b-qlora.yml`
- [x] tested with falcon 1b & 7b on A6000 (with `xformers_attention` turned off), training report: https://api.wandb.ai/links/utensil/y0iazatw 
- [x] change `max_packed_sequence_len` to empty 
- [x] use `gradient_accumulation_steps`
- [x] test falcon + qlora + xformer (training report updated)
- [x] test falcon 40b + qlora + xformer (training report updated)
- [x] Update qlora configs to the latest test ones, for now please see notebooks below ~or [this gist](https://gist.github.com/fearnworks/1163af11d4935faa6f81909f8a9321c0) verified by @fearnworks  (now merged)~
- [x] Isolate the qlora dependencies and produce a clean setup so everyone can use it
- Add templates and notebooks to jump start
  - [x] Colab notebook: https://colab.research.google.com/github/utensil/llm-playground/blob/main/notebooks/axolotl/colab/axolotl_falcon_1b_qlora_gsm8k.ipynb
  - [x] [RunPod minimal template](https://runpod.io/gsc?template=tkb65a1zcb&ref=km0th85l
) and notebook: https://github.com/utensil/llm-playground/blob/main/notebooks/axolotl/runpod/axolotl-falcon-1b-qlora-tutorial.ipynb
  - [x] LambdaLabs notebook: https://github.com/utensil/llm-playground/blob/main/notebooks/axolotl/lambdalabs/axolotl_falcon_1b_qlora_gsm8k.ipynb

- [ ] Try multi-GPU and compare training time to ft
- [ ] Evaluate inference result
- [ ] Errata section
- [ ] test falcon + qlora + flash attention [🧵 ](https://huggingface.co/tiiuae/falcon-40b/discussions/34)

To reproduce falcon + qlora:

- [The entrypoint script for the docker image](https://github.com/utensil/llm-playground/blob/main/scripts/entry/prepare_ax.sh) `winglian/axolotl-runpod:main-cu118-2.0.0`
- The training notebooks for [1b](https://github.com/utensil/llm-playground/blob/main/notebooks/axolotl/runpod/axolotl-falcon-1b-qlora-gsm8k.ipynb), [7b](https://github.com/utensil/llm-playground/blob/main/notebooks/axolotl/runpod/axolotl-falcon-7b-qlora-gsm8k.ipynb) and [40b](https://github.com/utensil/llm-playground/blob/main/notebooks/axolotl/runpod/axolotl-falcon-40b-qlora-gsm8k.ipynb) (check out file history for earlier training sessions, but not all runs are preserved)

Disclaimer: the config works, but might not be optimal. Improvements welcome!

### CODE REVIEW

**Comment by winglian** [comment]
Thanks! would you mind adding an Errata section to the bottom of the README, specifically that falcon/qlora/xformers doesn't work? someone will ultimately attempt to try that and just having it documented somewhere would be a great help.

**Comment by utensil** [comment]
OK, I'll test the combination today based on the new xformer patch landed in the docker image, and add the section. Hopefully I'll have an idea about why it won't work too.

I'll also test flash attention and change max packed sequence length to empty as suggested by caseus on Discord to see if it helps with the VRAM usage.

**Comment by utensil** [comment]
By an Errata section, do you mean that for each combination that axolotl doesn't support yet, there's a short description about the unsupported reason like errors and tracking issue?

Also, it would be nice to link check marks to example configs too, I was attempted to do so in the PR 😉

File: examples/falcon/config-7b-qlora.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+tokenizer_type: AutoTokenizer
+load_in_8bit: false
+load_in_4bit: true
+gptq: false
+strict: false
+push_dataset_to_hub:
+datasets:
+  - path: QingyiSi/Alpaca-CoT
+    data_files:
+      - Chain-of-Thought/formatted_cot_data/gsm8k_train.json
```
Comment (NanoCode012) [inline examples/falcon/config-7b-qlora.yml:18]: I think this should be removed since this is for your local setup. You may also remove the `data_files:` portion.
```suggestion
```

File: examples/falcon/config-7b-qlora.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+lora_model_dir:
+sequence_len: 2048
+max_packed_sequence_len: 2048
+lora_r: 64
+lora_alpha: 16
+lora_dropout: 0.05
+lora_target_modules:
+lora_target_linear: true
+lora_fan_in_fan_out:
+wandb_project: falcon-qlora
```
Comment (NanoCode012) [inline examples/falcon/config-7b-qlora.yml]:
```suggestion
wandb_project:
```

**Review by NanoCode012 — COMMENTED** [review]
Just some points I saw

File: examples/falcon/config-7b-qlora.yml (hunk 2 above)
Comment (utensil) [reply examples/falcon/config-7b-qlora.yml]: OK. Will do so after all the testing.

File: examples/falcon/config-7b-qlora.yml (hunk 1 above)
Comment (utensil) [reply examples/falcon/config-7b-qlora.yml:18]: I don't know if it still works after removing these lines. It's not a local setup but to specify fine-tuning only on this file in the dataset, since QingyiSi/Alpaca-CoT is a huge dataset collection.

Comment (utensil) [reply examples/falcon/config-7b-qlora.yml:14]: I don't know if it still works after removing these lines. It's not a local setup but to specify fine-tuning only on this file in the dataset, since QingyiSi/Alpaca-CoT is a huge dataset collection.

Comment (NanoCode012) [reply examples/falcon/config-7b-qlora.yml:18]: Thank you. I understand now. I had a misunderstanding of that config.

**Comment by FarisHijazi** [comment]
I just tried your changes, they don't work

Here's what I get:

```
You're using a PreTrainedTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/checkpoint.py:31: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")
Traceback (most recent call last):
  File "/workspace/axolotl/scripts/finetune.py", line 294, in <module>
    fire.Fire(train)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/workspace/axolotl/scripts/finetune.py", line 281, in train
    trainer.train(resume_from_checkpoint=resume_from_checkpoint)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1696, in train
    return inner_training_loop(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 1973, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2787, in training_step
    loss = self.compute_loss(model, inputs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/transformers/trainer.py", line 2819, in compute_loss
    outputs = model(**inputs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/peft/peft_model.py", line 663, in forward
    return self.base_model(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/root/.cache/huggingface/modules/transformers_modules/tiiuae/falcon-7b/da8d49a4c7dde3bfc39461e6f2cf7433e2fa44c2/modelling_RW.py", line 753, in forward
    transformer_outputs = self.transformer(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/root/.cache/huggingface/modules/transformers_modules/tiiuae/falcon-7b/da8d49a4c7dde3bfc39461e6f2cf7433e2fa44c2/modelling_RW.py", line 640, in forward
    outputs = torch.utils.checkpoint.checkpoint(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/checkpoint.py", line 249, in checkpoint
    return CheckpointFunction.apply(function, preserve, *args)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/utils/checkpoint.py", line 107, in forward
    outputs = run_function(*args)
  File "/root/.cache/huggingface/modules/transformers_modules/tiiuae/falcon-7b/da8d49a4c7dde3bfc39461e6f2cf7433e2fa44c2/modelling_RW.py", line 636, in custom_forward
    return module(*inputs, use_cache=use_cache, output_attentions=output_attentions)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/root/.cache/huggingface/modules/transformers_modules/tiiuae/falcon-7b/da8d49a4c7dde3bfc39461e6f2cf7433e2fa44c2/modelling_RW.py", line 385, in forward
    attn_outputs = self.self_attention(
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/root/.cache/huggingface/modules/transformers_modules/tiiuae/falcon-7b/da8d49a4c7dde3bfc39461e6f2cf7433e2fa44c2/modelling_RW.py", line 242, in forward
    fused_qkv = self.query_key_value(hidden_states)  # [batch_size, seq_length, 3 x hidden_size]
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/peft/tuners/lora.py", line 487, in forward
    result = F.linear(x, transpose(self.weight, self.fan_in_fan_out), bias=self.bias)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (392x4544 and 1x10614784)
```

File: examples/falcon/config-7b-qlora.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+wandb_watch:
+wandb_run_id:
+wandb_log_model:
+output_dir: ./qlora-out
+micro_batch_size: 40
+gradient_accumulation_steps: 2
+num_epochs: 3
+optimizer: paged_adamw_32bit
```
Comment (winglian) [inline examples/falcon/config-7b-qlora.yml:57]: does `paged_adamw_32bit` converge? I remember seeing some tests where this optimizer is problematic

Comment (utensil) [reply examples/falcon/config-7b-qlora.yml:57]: it seems fine in my tests, and it seems to help surviving VRAM spike

**Comment by utensil** [comment]
@FarisHijazi Hi, this happened before with outdated deps. In my tests, I updated to deps to master. I'll try to reproduce this in a raw setting and work my way back to determine which dep should be updated ASAP. For now, please check https://github.com/utensil/llm-playground/blob/main/scripts/prepare_qlora.sh .

**Comment by utensil** [comment]
I've just created [a minimal Colab notebook](https://colab.research.google.com/github/utensil/llm-playground/blob/main/notebooks/axolotl/colab/axolotl_falcon_1b_qlora_gsm8k.ipynb), so anyone can jump start to try it out. It's using a free T4 GPU instance.

The notebook works nice and clean as is.

But I'll break it down here to describe issues I encountered during creating that notebook, since others might encounter similar issues when deviating from the notebook:

## ModuleNotFoundError: No module named 'peft'

Fixed by `pip install git+https://github.com/huggingface/peft.git`

It should have been installed by axolotl but it's still not found.

Other packages in `4. Install QLoRA dependencies` are actually already installed and works fine.

UPDATE: Found the root cause of this and raised #151

## RuntimeError: self and mat2 must have the same dtype

This error appears if one encounters the error above, then run `pip install peft`. This would install an old version of peft before QLoRA PR landed in peft.

## ValueError: Your setup doesn't support bf16/gpu. You need torch>=1.10, using Ampere GPU with cuda>=11.0

This is a T4-specific issue.

Fixed by `bf16: false` . I like `bf16` whenever possible.

## ValueError: --tf32 requires Ampere or a newer GPU arch, cuda>=11 and torch>=1.7

Also a T4-specific issue.

Fixed by `tf32: false`. But I don't know its ramification.

**Comment by FarisHijazi** [comment]
I think most of these float type errors cna be solved by changing the gpu (colab peo or switching to another peovider)
I'll test it out
Btw i love your work

**Comment by utensil** [comment]
@FarisHijazi Thank you!

The root cause of the issue you encountered has been found. It is described in #151 . It's easy to fix it by running `pip install -U git+https://github.com/huggingface/peft.git` .

**Comment by FarisHijazi** [comment]
damn..... shape error getting solved by upgrading a dependency...
insane

will try to run it now

**Comment by FarisHijazi** [comment]
getting this error on my machine, will try some other things and update you
```
GPT2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
Error invalid device ordinal at line 359 in file /home/tim/git/bitsandbytes/csrc/pythonInterface.c
/arrow/cpp/src/arrow/filesystem/s3fs.cc:2598:  arrow::fs::FinalizeS3 was not called even though S3 was initialized.  This could lead to a segmentation fault at exit
Traceback (most recent call last):
  File "/root/miniconda3/envs/py3.9/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/accelerate/commands/accelerate_cli.py", line 45, in main
    args.func(args)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/accelerate/commands/launch.py", line 918, in launch_command
    simple_launcher(args)
  File "/root/miniconda3/envs/py3.9/lib/python3.9/site-packages/accelerate/commands/launch.py", line 580, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['/root/miniconda3/envs/py3.9/bin/python3', 'scripts/finetune.py', 'examples/falcon/config-1b-qlora-utensil.yml']' returned non-zero e
```

**Comment by utensil** [comment]
@FarisHijazi Possible fix to that: https://github.com/TimDettmers/bitsandbytes/issues/425#issuecomment-1562283082 

Can you elaborate on the environment or setup? (Maybe on a separate issue to better follow up)

**Comment by FarisHijazi** [comment]
Will try to get on that when i finish work

Btw i do get a bitsandbytes warning when i finetune the other models but it does train. But for falcon i get a warning and then later an error
Not sure if that's a problem
Will post later when I'm on my pc

**Comment by NanoCode012** [comment]
Hello @utensil , thank you for the extensive work and report. 

Since, you have tested it working for falcon+qlora, would you be interested in merging this first? I also saw that you have included ipynb. Do you want to also upload the notebook to the same folder as the config before merge?

Regarding winglian's initial comment, he might've meant to add a short description for that combination here: https://github.com/OpenAccess-AI-Collective/axolotl#common-errors-

Lastly, you can follow up with a separate Issue for your other task (multi-gpu etc)?

**Comment by utensil** [comment]
Hi @NanoCode012 , thank you for the review, I'm OK to merge it as is, and leave other tasks including the notebook tasks and the errata section (I'm not so sure if they're still there now) to future PRs.

**Comment by NanoCode012** [comment]
Would it be possible to fix the failing `pre-commit`? You can see the error by pressing Details here or running 

```bash
pre-commit install # if haven't before

pre-commit run --all-files
```

File: examples/falcon/config-7b-qlora.yml [hunk 4]
```diff
@@ -0,0 +1,92 @@
+# 1b: tiiuae/falcon-rw-1b
+# 40b: tiiuae/falcon-40b
+base_model: tiiuae/falcon-7b
+base_model_config: tiiuae/falcon-7b
+# required by falcon custom model code: https://huggingface.co/tiiuae/falcon-7b/tree/main
```
Comment (utensil) [inline examples/falcon/config-7b-qlora.yml]:
```suggestion
# required by falcon custom model code: https://huggingface.co/tiiuae/falcon-7b/tree/main
```

**Comment by NanoCode012** [comment]
Thank you for the amazing work!

**Comment by ehartford** [comment]
nice I gotta try this with Samantha
